lib/upgrade/etcd/conversionv1v3/workloadendpoint.go

```diff
@@ -0,0 +1,276 @@
+			Endpoint:      convertName(wepKey.EndpointID),
+			IPNetworks:    ipNets,
+			IPNATs:        ipNats,
+			Profiles:      convertProfiles(wepValue.ProfileIDs),
+			InterfaceName: wepValue.Name,
+			MAC:           wepValue.Mac.String(),
+			Ports:         convertPorts(wepValue.Ports),
+		},
+	}
+
+	if wepValue.IPv4Gateway != nil {
+		wep.Spec.IPv4Gateway = wepValue.IPv4Gateway.String()
+	}
+	if wepValue.IPv6Gateway != nil {
+		wep.Spec.IPv6Gateway = wepValue.IPv6Gateway.String()
+	}
+
+	// Figure out the new name based on WEP fields.
+	wepids := names.WorkloadEndpointIdentifiers{
+		Node:         wep.Spec.Node,
+		Orchestrator: wep.Spec.Orchestrator,
+		Endpoint:     wep.Spec.Endpoint,
+		Workload:     wep.Spec.Workload,
+		Pod:          wep.Spec.Pod,
+		ContainerID:  wep.Spec.ContainerID,
+	}
+
+	name, err := wepids.CalculateWorkloadEndpointName(false)
+	if err != nil {
+		return nil, err
+	}
+	wep.ObjectMeta.Name = name
+
+	log.Debugf("Converted: %+v\n To: %+v", kvp, wep)
+
+	return &wep, nil
+}
+
+func convertMetadataToKey(m unversioned.ResourceMetadata) (model.Key, error) {
+	hm := m.(apiv1.WorkloadEndpointMetadata)
+	k := model.WorkloadEndpointKey{
+		Hostname:       hm.Node,
+		OrchestratorID: hm.Orchestrator,
+		WorkloadID:     hm.Workload,
+		EndpointID:     hm.Name,
+	}
+	return k, nil
+}
+
+// convertLabels creates a new map of labels, and updates the v1 namespace label to the v3 style.
+func convertLabels(v1Labels map[string]string) map[string]string {
+	labels := map[string]string{}
+	for k, v := range v1Labels {
+		labels[k] = v
+	}
+
+	if val, ok := labels["calico/k8s_ns"]; ok {
+		labels["projectcalico.org/namespace"] = val
+		delete(labels, "calico/k8s_ns")
+	}
+
+	return labels
+}
+
+// convertIPNetworks updates the old []net.IPNet to []string.
+func convertIPNetworks(ipNetworks []net.IPNet) []string {
+	var ipNets []string
+	for _, ipNet := range ipNetworks {
+		ipNets = append(ipNets, ipNet.String())
+	}
+
+	return ipNets
+}
+
+// convertIPNATs updates the type of IPNAT struct used.
+func convertIPNATs(v1IPNATs []model.IPNAT) []apiv3.IPNAT {
+	var ipNATs []apiv3.IPNAT
+	for _, ipNAT := range v1IPNATs {
+		ipNATs = append(ipNATs, apiv3.IPNAT{
+			InternalIP: ipNAT.IntIP.String(),
+			ExternalIP: ipNAT.ExtIP.String(),
+		})
+	}
+
+	return ipNATs
+}
+
+// convertProfiles updates the Kubernetes namespace portion from "k8s_ns" to "kns" for each profile.
+func convertProfiles(v1Profiles []string) []string {
+	var v3Profiles []string
+	for _, p := range v1Profiles {
+		if strings.HasPrefix(p, "k8s_ns.") {
+			parts := strings.Split(p, ".")
+			prof := "kns." + strings.Join(parts[1:], "")
+			v3Profiles = append(v3Profiles, prof)
+		} else {
+			v3Profiles = append(v3Profiles, p)
+		}
+	}
+
+	return v3Profiles
+}
+
+// getPodName pulls the pod name out for k8s based installs, the pod name is mixed in with the
+// WorkloadID as `namespace.pod_name` so we separate them out.
+func getPodName(workload string, namespace string) string {
```

Comment:
I think this should be `getPodNamespaceName` and it should return both, rather than take the namespace as an argument.

Just split after the first dot.